libraries/aws/provisioning/src/aws_iot_provisioning_api.c

```diff
@@ -517,6 +533,195 @@ AwsIotProvisioningError_t AwsIotProvisioning_CreateKeysAndCertificate( IotMqttCo
+        .pTopicFilterBase      = responseTopicsBuffer,
+        .topicFilterBaseLength = PROVISIONING_CREATE_CERT_FROM_CSR_RESPONSE_TOPIC_FILTER_LENGTH
+    };
+    size_t payloadSize = 0;
+    uint8_t * pPayloadBuffer = NULL;
+    IotMqttPublishInfo_t publishInfo = IOT_MQTT_PUBLISH_INFO_INITIALIZER;
+
+    AwsIotProvisioningError_t status = AWS_IOT_PROVISIONING_SUCCESS;
+
+    /* Verify that library has been initialized. */
+    AwsIotProvisioning_Assert( _checkInit() == true );
+
+    if( connection == IOT_MQTT_CONNECTION_INITIALIZER )
+    {
+        IotLogError( "Bad parameter: MQTT connection is not initialized: Operation={%s}",
+                     CREATE_CERT_FROM_CSR_OPERATION_LOG );
+
+        status = AWS_IOT_PROVISIONING_BAD_PARAMETER;
+    }
+    else if( ( pCertificateSigningRequest == NULL ) || ( csrLength == 0 ) )
+    {
+        IotLogError( "Bad parameter: Invalid Certificate-Signing Request: "
+                     "Operation={%s}", CREATE_CERT_FROM_CSR_OPERATION_LOG );
+
+        status = AWS_IOT_PROVISIONING_BAD_PARAMETER;
+    }
+    /* Check that a callback function object along with a valid callback functor is provided. */
+    else if( ( pResponseCallback == NULL ) ||
+             ( pResponseCallback->function == NULL ) )
+    {
+        IotLogError(
+            "Bad parameter: Invalid callback provided: Operation={%s}",
+            CREATE_CERT_FROM_CSR_OPERATION_LOG );
+
+        status = AWS_IOT_PROVISIONING_BAD_PARAMETER;
+    }
+
+    if( status == AWS_IOT_PROVISIONING_SUCCESS )
+    {
+        /* Copy the response topics in a local buffer for appropriate suffixes to be added. */
+        ( void ) memcpy( responseTopicsBuffer, PROVISIONING_CREATE_KEYS_AND_CERTIFICATE_RESPONSE_TOPIC_FILTER,
+                         PROVISIONING_CREATE_KEYS_AND_CERTIFICATE_RESPONSE_TOPIC_FILTER_LENGTH );
+
+        /* Subscribe to the MQTT response topics. */
+        mqttOpResult = AwsIot_ModifySubscriptions( IotMqtt_SubscribeSync, &responseSubscription );
+
+        if( mqttOpResult != IOT_MQTT_SUCCESS )
+        {
+            IotLogError( "Failed to subscribe to response topics: Operation={%s}, MQTTError={%s}",
+                         CREATE_CERT_FROM_CSR_OPERATION_LOG,
+                         IotMqtt_strerror( mqttOpResult ) );
+            status = AWS_IOT_PROVISIONING_MQTT_ERROR;
+        }
+    }
+
+    if( status == AWS_IOT_PROVISIONING_SUCCESS )
+    {
+        IotLogDebug( "Subscribed to response topics: Operation={%s}",
+                     CREATE_CERT_FROM_CSR_OPERATION_LOG );
+
+        /* Update the operation object to represent an active "Certificate-Signing Request" operation. */
+        _provisioningCallbackInfo_t callbackInfo;
+        callbackInfo.createCertificateFromCsrCallback = *pResponseCallback;
+        _setActiveOperation( &callbackInfo );
+
+        /* Serialization of request payload occurs in a 2-step process, one for calculation of buffer size, and then,
```

Comment:
This is a common idiom, of course. There is a cost, though, which is that computing the payload length may require the whole request to be serialized and then thrown away.

An alternative approach is to pass the allocator into the function.

Reply:
As the `tinycbor` library we are using, requires the buffer to be pre-allocated, we cannot avoid the 2-step process of serialization in any layer. The same is followed by any other library performing CBOR encoding (Defender, BLE (I think))
For this (CSR) feature's serializer, I deviated from the existing serializer functions (of the other APIs in the library) to make it the responsibility of the API function/layer to allocate the memory instead of making the serializer do so to align follow the principle of single-responsibility.

Reply:
Is the callee having to do the encoding twice in order to service the size-only call? Or can it compute the required size just based on the input size?

Reply:
Yes, the callee is effectively doing the encoding twice. As mentioned in a previous comment-chain reply, I have broken the serializer function into 2 for better code readability